content/chapters/app-interact/lecture/slides.mdpp

```diff
@@ -0,0 +1,214 @@
+
+```c
+// writer.c
+void main()
+{
+    FILE *fp = fopen("myf.txt", "r");
+    fprintf(fp, "Hello");
+    fclose(fp);
+}
+```
+```c
+// reader.c
+void main()
+{
+    char a[20];
+    FILE *fp = fopen("myf.txt", "r");
+    fscanf(fp, "%s", a);
+    printf("%s\n", a);
+    fclose(fp);
+}
+```
+----
+### Exhibit 2
+
+* The 2 programs indeed interact.
+* But is there any difference from the previous example?
+* There is no protocol of interaction.
+* If one of the applications did not exist, the other one could still be a valid, meaningful program.
+----
+
+### Exhibit 3
+
+```c
+int main ()
+{
+    int pid = fork();
+    if (pid == 0)
+    {
+        sleep(2);
+        exit(EXIT_SUCCESS);
+    }
+
+    waitpid(pid, &status, 0);
+    return 0;
+}
+```
+----
+
+### Exhibit 3
+
+* The programs interact in a structured manner.
+* One process waits for the other to finish.
+* Interaction is dynamic - e.g. a process actively waits for the other.
+* The 2 processes have been designed to work toghether.
+----
+
+### Exhibit 4 - WhatsApp application
+
+![App1](media/whatsApp.svg)
+----
+
+### Exhibit 4 - WhatsApp application
+
+* There is interaction between the main client application and other applications: photos, location, youtube etc.
+* There is also interaction with the web server when sending messages.
+
+----
+
+### What does it mean that applications interact?
+
+* Applications that run at the same time and that communicate with each other during their lifetimes.
+* This communication involves passing messages and waiting for responses.
+
+----
+
+* This interaction is done between coupled or decoupled applications
+* We distinguish between:
+    - interaction of applications on the same system
+    - interaction of applications on different systems
+* Applications may communicate
+
+----
+
+## How do Apps Interact?
+
+* An application that strives to communicate with other applications must offer an interface.
+* The application may offer a user interface.
+* TODO: image with user-app interface, app-app interface (on same system and on remote systems)
+
+----
+
+## OS - The Lawmaker
+
+* The OS offers the primitives for app-app intercation
+* The OS is like a lawmaker the dictates how the interaction takes place
+* TODO: image with user-app, app-app interface on top of OS (same/remote system)
+
+---
+
+## App Classification
+
+- Uni-process (single/multi threaded)
+- Multi-process
+- Multi-threaded
+- Multi-system
+- Examples
+- These apps use OS primitives to communicate between them
+
+---
+
+## Interaction Models
+
+- Inter-application
+- Intra-application
+- Communication channel
+- Shared memory
+
+----
+
+## Communication channel
+
+- TODO: img
+- messages, packets
+- sender receiver
+- OS does the transmitting
+- requires communication protocol
+- used for inter-application interaction, but could be used for intra-app communication also
+
+----
+
+## Shared memory
+
+- TODO: img
+- apps have a common memory region they use to communicate
```

Comment:
Be more specific that different processes map the same frames. You can use [this demo](https://github.com/open-education-hub/operating-systems/blob/master/content/chapters/compute/lab/content/arena.md#shared-memory) as a starting point.